File: main/src/render/Ray.cs
using System.Runtime.InteropServices;
using Silk.NET.Maths;

[StructLayout(LayoutKind.Sequential)]
public struct Ray {

    public float OriginX;
    public float OriginY;
    public float OriginZ;

    public float DirectionX;
    public float DirectionY;
    public float DirectionZ;

    public static Ray FromVectors(Vector3D<float> origin, Vector3D<float> direction) => new Ray {
        OriginX = origin.X,
        OriginY = origin.Y,
        OriginZ = origin.Z,
        DirectionX = direction.X,
        DirectionY = direction.Y,
        DirectionZ = direction.Z
    };

}

[StructLayout(LayoutKind.Sequential)]
public struct RaycastResult {
    public byte Hit;
    public int X;
    public int Y;
    public int Z;
    public byte Face;
}

File: brain/src/ray.rs
#[repr(C)]
pub struct Ray {
    pub origin_x: f32,
    pub origin_y: f32,
    pub origin_z: f32,
    pub direction_x: f32,
    pub direction_y: f32,
    pub direction_z: f32,
}

#[repr(C)]
pub struct RaycastResult {
    pub hit: u8,
    pub x: i32,
    pub y: i32,
    pub z: i32,
    pub face: u8,
}
